### PR TITLE
Use version classes for perf

### PR DESCRIPTION
**Pull request by atuchin-m**
Resolves https://github.com/brave/brave-browser/issues/28192

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: tools/perf/components/browser_type.py [hunk 1]
```diff
@@ -19,11 +19,57 @@
...
+  def __init__(self, tag: str) -> None:
+    super().__init__()
+    assert re.match(r'v\d+\.\d+\.\d+', tag)
+    self._version = list(map(int, tag[1:].split('.')))
+
+  def __str__(self) -> str:
+    return 'v' + '.'.join(map(str, self._version))
+
+  def to_chromium_version(self) -> 'ChromiumVersion':
+    return _GetChromiumVersion(self)
```
Comment (goodov) [inline tools/perf/components/browser_type.py:50]: is `git show` quick enough? how many calls to this method? maybe we should have a cached map `brave_version->chromium.major`? or at least cache what `_GetChromiumVersion` returns if we only work with a single `BraveVersion` during the run.

File: tools/perf/components/perf_test_runner.py [hunk 2]
```diff
@@ -291,12 +299,14 @@ def PrepareBinariesAndDirectories(configurations: List[RunnerConfig],
...
 
     binary_path = None
+    field_trial_config = None
 
     if common_options.do_run_tests:
       shutil.rmtree(out_dir, True)
       os.makedirs(out_dir)
       binary_path = PrepareBinary(out_dir, config.tag, config.location,
                                   config.browser_type)
+      assert (config.tag is not None)
```
Comment (goodov) [inline tools/perf/components/perf_test_runner.py:309]: shouldn't it be `assert config.tag is not None`?

File: tools/perf/components/browser_type.py [hunk 3]
```diff
@@ -19,11 +19,57 @@
...
 
 
-def _GetBraveDownloadUrl(tag: str, binary: str) -> str:
+class _BaseVersion:
+  _version: List[int]
+
+  def equals(self, a: _BaseVersion) -> bool:
+    return self._version == a.version()
+
+  def less(self, a: _BaseVersion) -> bool:
```
Comment (goodov) [inline tools/perf/components/browser_type.py:32]: why not `def __eq__(self, other):` and `def __lt__(self, other):`?

File: tools/perf/components/browser_binary_fetcher.py [hunk 4]
```diff
@@ -8,19 +8,21 @@
...
-def ParseTarget(target: str) -> Tuple[Optional[str], str]:
+def ParseTarget(target: str) -> Tuple[Optional[BraveVersion], str]:
   m = re.match(r'^(v\d+\.\d+\.\d+)(?::(.+)|$)', target)
   if not m:
     return None, target
-  logging.debug('Parsed tag: %s, location : %s', m.group(1), m.group(2))
-  return m.group(1), m.group(2)
+  tag = BraveVersion(m.group(1))
+  localtion = m.group(2)
+  logging.debug('Parsed tag: %s, location : %s', tag, localtion)
```
[the remark this answers is not in the document]
Comment (atuchin-m) [reply tools/perf/components/browser_binary_fetcher.py:20]: https://pylint.pycqa.org/en/latest/user_guide/messages/warning/logging-fstring-interpolation.html

File: tools/perf/components/browser_type.py (hunk 1 above)
Comment (atuchin-m) [reply tools/perf/components/browser_type.py:50]: We call `git something` a few times per each test binary. It adds about 10sec to the test binary time. In total we have about 30min, so it doesn't look like a place to optimize now. 

After an integration with CI some git operations will be removed, so we can think about caching that time.

File: tools/perf/components/browser_binary_fetcher.py [hunk 5]
```diff
@@ -8,19 +8,21 @@
...
 
-def ParseTarget(target: str) -> Tuple[Optional[str], str]:
+def ParseTarget(target: str) -> Tuple[Optional[BraveVersion], str]:
   m = re.match(r'^(v\d+\.\d+\.\d+)(?::(.+)|$)', target)
   if not m:
     return None, target
-  logging.debug('Parsed tag: %s, location : %s', m.group(1), m.group(2))
-  return m.group(1), m.group(2)
+  tag = BraveVersion(m.group(1))
+  localtion = m.group(2)
```
[the remark this answers is not in the document]
Comment (atuchin-m) [reply tools/perf/components/browser_binary_fetcher.py]: Fixed

File: tools/perf/components/perf_test_runner.py (hunk 2 above)
Comment (atuchin-m) [reply tools/perf/components/perf_test_runner.py]: Fixed

File: tools/perf/components/browser_type.py (hunk 3 above)
Comment (atuchin-m) [reply tools/perf/components/browser_type.py]: Fixed

**Review by goodov — APPROVED** [review]

**Pull request merged**